test/cpp/end2end/spiffe_bundle_map_end2end_test.cc

```diff
@@ -0,0 +1,474 @@
+              kServerSpiffeBundleMapPath);
+  });
+  notification.WaitForNotification();
+
+  auto certificate_provider =
+      std::make_shared<experimental::FileWatcherCertificateProvider>(
+          std::string(kClientKeyPath), std::string(kClientCertPath),
+          /*root_cert_path=*/"", tmp_bundle_map.name(), 1);
+  grpc::experimental::TlsChannelCredentialsOptions options;
+  options.set_certificate_provider(certificate_provider);
+  options.watch_root_certs();
+  options.set_root_cert_name("root");
+  options.watch_identity_key_cert_pairs();
+  options.set_identity_cert_name("identity");
+
+  options.set_check_call_host(false);
+  auto verifier = std::make_shared<experimental::NoOpCertificateVerifier>();
+  options.set_certificate_verifier(verifier);
+  DoRpc(server_addr_, options, true);
+
+  // Update the spiffe bundle map to something that will fail
+  tmp_bundle_map.RewriteFile(server_bundle_map);
+  // Wait 2 seconds to ensure a refresh happens
+  gpr_sleep_until(gpr_time_add(gpr_now(GPR_CLOCK_MONOTONIC),
+                               gpr_time_from_seconds(2, GPR_TIMESPAN)));
+  constexpr absl::string_view expected_message_start =
+      "failed to connect to all addresses; last error: ";
+  const StatusCode expected_status = StatusCode::UNAVAILABLE;
+  DoRpc(server_addr_, options, false,
+        MakeTlsHandshakeFailureRegex(expected_message_start), expected_status);
+}
+
+TEST_F(SpiffeBundleMapTest, ServerBadMap) {
+  server_addr_ = absl::StrCat("localhost:",
+                              std::to_string(grpc_pick_unused_port_or_die()));
+  absl::Notification notification;
+  // Use the client-side spiffe bundle map on the server side to force a failure
+  server_thread_ = new std::thread([&]() {
+    RunServer(&notification, kServerKeyPath, kServerCertPath, "",
+              kClientSpiffeBundleMapPath);
+  });
+  notification.WaitForNotification();
+
+  auto certificate_provider =
+      std::make_shared<experimental::FileWatcherCertificateProvider>(
+          std::string(kClientKeyPath), std::string(kClientCertPath),
+          /*root_cert_path=*/"", std::string(kClientSpiffeBundleMapPath), 1);
+  grpc::experimental::TlsChannelCredentialsOptions options;
+  options.set_certificate_provider(certificate_provider);
+  options.watch_root_certs();
+  options.set_root_cert_name("root");
+  options.watch_identity_key_cert_pairs();
+  options.set_identity_cert_name("identity");
+
+  options.set_check_call_host(false);
+  auto verifier = std::make_shared<experimental::NoOpCertificateVerifier>();
+  options.set_certificate_verifier(verifier);
+
+  constexpr absl::string_view expected_message_start =
+      "failed to connect to all addresses; last error: ";
+  const StatusCode expected_status = StatusCode::UNAVAILABLE;
+  DoRpc(server_addr_, options, false,
+        MakeConnectionFailureRegex(expected_message_start), expected_status);
+}
+
+TEST_F(SpiffeBundleMapTest, ClientBadMap) {
```

Comment:
This test name is ambiguous, suggest renaming to something more descriptive, e.g. "ClientSideSpiffeVerificationFailure".